### ISSUE TITLE
Verify role against a list of allowed roles

### ISSUE THREAD

**Issue by marcysutton** [angular/angular#754]
As discussed in #630, it would be hugely valuable to validate role values against a list of allowed roles. It is a relatively fixed list–the ones we'd want to allow include widget, document, and landmark roles. Abstract roles should throw a warning along with invalid ones. 

Some examples that should throw errors:
- `role="image"` – Error, not a valid role. Should be `img`.
- `role="input"` – Error, abstract roles cannot be used directly. Should be `textbox`.

Specifying valid replacements like `img` instead of `image` would add complexity but would be super helpful.

<h2>Roles to allow:</h2>

<h3>Widgets</h3>
- <a href="http://www.w3.org/TR/wai-aria/roles#alert"><code>alert</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#alertdialog"><code>alertdialog</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#button"><code>button</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#checkbox"><code>checkbox</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#dialog"><code>dialog</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#gridcell"><code>gridcell</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#link"><code>link</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#log"><code>log</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#marquee"><code>marquee</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#menuitem"><code>menuitem</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#menuitemcheckbox"><code>menuitemcheckbox</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#menuitemradio"><code>menuitemradio</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#option"><code>option</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#progressbar"><code>progressbar</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#radio"><code>radio</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#scrollbar"><code>scrollbar</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#slider"><code>slider</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#spinbutton"><code>spinbutton</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#status"><code>status</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#tab"><code>tab</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#tabpanel"><code>tabpanel</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#textbox"><code>textbox</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#timer"><code>timer</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#tooltip"><code>tooltip</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#treeitem"><code>treeitem</code></a>

<h3>Composite Widgets</h3>
- <a href="http://www.w3.org/TR/wai-aria/roles#combobox"><code>combobox</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#grid"><code>grid</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#listbox"><code>listbox</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#menu"><code>menu</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#menubar"><code>menubar</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#radiogroup"><code>radiogroup</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#tablist"><code>tablist</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#tree"><code>tree</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#treegrid"><code>treegrid</code></a>

<h3>Document Structure</h3>
- <a href="http://www.w3.org/TR/wai-aria/roles#article"><code>article</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#columnheader"><code>columnheader</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#definition"><code>definition</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#directory"><code>directory</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#document"><code>document</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#group"><code>group</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#heading"><code>heading</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#img"><code>img</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#list"><code>list</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#listitem"><code>listitem</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#math"><code>math</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#note"><code>note</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#presentation"><code>presentation</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#region"><code>region</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#row"><code>row</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#rowgroup"><code>rowgroup</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#rowheader"><code>rowheader</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#separator"><code>separator</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#toolbar"><code>toolbar</code></a>

<h3>Landmark Roles</h3>
- <a href="http://www.w3.org/TR/wai-aria/roles#application"><code>application</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#banner"><code>banner</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#complementary"><code>complementary</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#contentinfo"><code>contentinfo</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#form"><code>form</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#main"><code>main</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#navigation"><code>navigation</code></a>
- <a href="http://www.w3.org/TR/wai-aria/roles#search"><code>search</code></a>

**Comment by dylanb** [comment]
In order to avoid difficult to locate problems, Angular.js should

1) validate that the role does not get changed after it has been set (preferred), or when this is done, create an entirely new node and delete and replace the old one (see spec below). Ideally, setting the role should only be done before the node gets inserted into the DOM (or before the current JavaScript execution ends) because there are implicit roles associated with elements that might get picked up immediately when the node gets inserted.
2) validate required parent and child roles. For example menuitem cannot exist without a parent menu role.
3) stretch goal: validate interim structures that need role="presentation" in order for the AT to handle them correctly

For 1 above, quoting from the spec:

Roles are element types and authors MUST NOT change role values over time or with user actions. Authors wishing to change a role MUST do so by deleting the associated element and its children and replacing it with a new element with the appropriate role.

http://www.w3.org/TR/wai-aria/roles

For 3 above - the following is invalid menu markup

```html
<ul role="menubar">
    <li role="menuitem">
        Deals
        <ul role="menu">
            <li>
                <a href="vegas" role="menuitem">
                    Vegas
                </a>
            </li>
            <li>
                <a href="caracas" role="menuitem">
                    Caracas
                </a>
            </li>
        </ul>
    </li>
</ul>
```

Whereas this is valid markup because all interim structures in the hierarchy have been annotated with role="presentation"

```html
<ul role="menubar">
    <li role="menuitem">
        Deals
        <ul role="menu">
            <li role="presentation">
                <a href="vegas" role="menuitem">
                    Vegas
                </a>
            </li>
            <li role="presentation">
                <a href="caracas" role="menuitem">
                    Caracas
                </a>
            </li>
        </ul>
    </li>
</ul>
```

**Comment by splincode** [comment]
@marcysutton Is the issue relevant?

**Comment by marcysutton** [comment]
@splincode at this point it's probably easier to integrate a tool like [axe-core](https://github.com/dequelabs/axe-core) and its existing ARIA infrastructure into the Angular CLI or build system than to maintain a list of allowed roles in Angular. It's a question of whether this should be done at runtime or build time....I wouldn't think Angular would want to ship all the code for accessibility testing over the wire, so build time is probably better for accessibility testing. It could be run on save, in the test suite and in CI. I have some materials on doing that: http://marcysutton.github.io/a11y-testing-with-angular/#/

There's also an issue open to integrate axe-core into Angular.io: https://github.com/angular/angular/issues/21969

**Comment by petebacondarwin** [comment]
Following on from @marcysutton's comment above, this kind of checking should probably be done by a tool like Axe.
